# Session 2: Genomics Techniques
This session will provide an overview for understanding FASTQ file formats, interpreting read quality, mapping to a reference, and calling genomic variants. It is derived from the [Data Wrangling and Processing for Genomics Lesson](https://datacarpentry.org/wrangling-genomics/) from [Data Carpentry](https://datacarpentry.org/lessons/) which provides a much better and in-depth view of these topics.  

To facilitate analysis on your local machine, a smaller dataset (instead of the <i>E. coli</i> used for the Data Carpentry lession) is provided. 

## Prior to the session: 
 
### For those working on Mac/Linux machines:

### For those working on Windows machines:
- Cygwin 
- Fastqc
- BWA, samtools, and bcftools software is unfortunately not available for Windows machines. Participants are welcome to follow along virtually during the workshop.  For the tenacious participant wishing to perform these analyses on a Windows machine, it may be worth looking into [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10).  You can choose which Linux distribution to install; Ubuntu 18.04 LTS is recommended.

